fix(settings): Heal invalid ToxPk saved to Personal Settings

anthonybilinski opened the pull request:
- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/6507)
<!-- Reviewable:end -->

anthonybilinski committed:
fix(settings): Heal invalid ToxPk saved to Personal Settings

Fix #6506

sudden6 commented:
Unfortunately this doesn't seem to fix https://github.com/qTox/qTox/issues/6506 for me.

New backtrace:
```
#0  0x00007ffff0bdecdb in raise () from /lib64/libc.so.6
#1  0x00007ffff0be0375 in abort () from /lib64/libc.so.6
#2  0x00007ffff0bd6cda in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ffff0bd6d62 in __assert_fail () from /lib64/libc.so.6
#4  0x00000000004fbacd in ToxPk::<lambda()>::operator()(void) const (__closure=0x7fffffffc340) at ../src/core/toxpk.cpp:86
#5  0x00000000004fbb13 in ToxPk::ToxPk (this=0x7fffffffc400, pk=...) at ../src/core/toxpk.cpp:89
#6  0x000000000058393a in Settings::loadPersonal (this=0x1612860, profileName=..., passKey=0x0) at ../src/persistence/settings.cpp:552
#7  0x0000000000580f89 in Settings::updateProfileData (this=0x1612860, profile=0x16c0030, parser=0x7fffffffd9d8) at ../src/persistence/settings.cpp:295
#8  0x000000000057228b in Profile::loadProfile (name=..., password=..., settings=..., parser=0x7fffffffd9d8) at ../src/persistence/profile.cpp:339
#9  0x0000000000531877 in Nexus::onLoadProfile (this=0x1612a70, name=..., pass=...) at ../src/nexus.cpp:311
```

sudden6 commented on src/persistence/settings.cpp:
```diff
@@ -466,16 +489,37 @@ void Settings::loadPersonal(QString profileName, const ToxEncrypt* passKey)
-    if (QFile(tmp).exists()) // otherwise, filePath remains the global file
+    if (QFile(tmp).exists()) { // otherwise, filePath remains the global file
         filePath = tmp;
+        defaultSettings = true;
```
This detection is buggy, when the file exists we can determine its version from the content.

anthonybilinski replied:
Thanks for finding that, updated.

sudden6 replied:
The update seems to have fixed it.

sudden6 reviewed (changes requested):
> Reviewed 1 of 4 files at r8, 60 of 60 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6507)* status: 0 of 1 LGTMs obtained

___
*[src/core/toxpk.h, line 27 at r9](https://reviewable.io/reviews/qtox/qtox/6507#-My1zwV1CGpKhcvAMl4I:-My1zwV1CGpKhcvAMl4J:brdmh1v) ([raw file](https://github.com/qtox/qtox/blob/ee63a43b69cd36a7f15326b768d0112aa97bac79/src/core/toxpk.h#L27)):*
> ```objc
> #include <cstdint>
> 
> #define PUBLIC_KEY_HEX_CHARS (2 * TOX_PUBLIC_KEY_SIZE)
> ```

I don't like defines in C++ if they can be avoided. Additonally, this makes everything that uses the `ToxPk` transitively depend on tox.h, can you find a better solution?


<!-- Sent from Reviewable.io -->

anthonybilinski committed:
refactor(ToxPk): Don't define model's ToxPk, ToxId, GroupId from tox.h

Since they exist in saved settings and the chat database, they are their own
size. assert that it matches tox.h's size in Core constructor, though.

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6507)* status: 0 of 1 LGTMs obtained

___
*[src/core/toxpk.h, line 27 at r9](https://reviewable.io/reviews/qtox/qtox/6507#-My1zwV1CGpKhcvAMl4I:-My2FNzfDqo8vdnLuZYd:bszjvw0) ([raw file](https://github.com/qtox/qtox/blob/ee63a43b69cd36a7f15326b768d0112aa97bac79/src/core/toxpk.h#L27)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

I don't like defines in C++ if they can be avoided. Additonally, this makes everything that uses the `ToxPk` transitively depend on tox.h, can you find a better solution?
</blockquote></details>

Removed define, and removed dependence on tox.h entirely. ToxPk, ToxId, and GroupId now define their lengths themselves rather than relying on tox.h even internally. I think this is better since they're used in our serialization in Settings and History, so their model extends beyond toxcore. I added asserts in the core constructor that the sizes match though.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 15 of 15 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6507#-:-My2MibN3OhBUVvQOyRC:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->
